espnet2/speechlm/core_lm/ar.py

```diff
@@ -0,0 +1,190 @@
+        att_unit: int = 256,
+        head: int = 2,
+        unit: int = 1024,
+        layer: int = 4,
+        dropout_rate: float = 0.0,
+        positional_dropout_rate: float = 0.0,
+        attention_dropout_rate: float = 0.0,
+    ):
+        super(ARLM, self).__init__()
+
+        if pos_enc == "sinusoidal":
+            pos_enc_class = PositionalEncoding
+        else:
+            raise ValueError(f"unknown pos-enc option: {pos_enc}")
+
+        self.emb = torch.nn.Embedding(vocab_size, att_unit)
+        self.lm_head = torch.nn.Linear(att_unit, vocab_size * nq, bias=False)
+        if share_emb:
+            raise ValueError("Embedding cannot be shared")
+
+        self.decoders = torch.nn.ModuleList(
+            [
+                TransformerLayer(
+                    att_unit=att_unit,
+                    head=head,
+                    unit=unit,
+                    dropout_rate=dropout_rate,
+                    attention_dropout_rate=attention_dropout_rate,
+                    causal=True,
+                    cross_attention=False,
+                )
+                for _ in range(layer)
+            ]
+        )
+        self.post_ln = torch.nn.LayerNorm(att_unit)
+        self.pos_enc = pos_enc_class(att_unit, positional_dropout_rate)
+
+        self.nq = nq
+
+    def forward(
+        self,
+        dec_seq: torch.Tensor,
+        dec_seq_lengths: torch.Tensor = None,
+        enc_seq: torch.Tensor = None,
+        enc_seq_lengths: torch.Tensor = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor, Dict]:
+
+        assert dec_seq.dim() == 3
+
+        x = dec_seq[:, :-1]
+        logits = self.encode(x)
+
+        target = dec_seq[:, 1:]
+        loss, stats, weight = ce_loss(logits, target, dec_seq_lengths - 1)
+
+        return loss, stats, weight
+
+    def encode(self, x: torch.Tensor, cache: dict={}):
+        x = self.emb(x).sum(dim=2)
+
+        x = self.pos_enc(x, cache)
+        for layer in self.decoders:
+            x = layer(x, cache=cache)
+        x = self.post_ln(x)
+
+        logits = self.lm_head(x)
+        B, T, Vnq = logits.size()
+        logits = logits.view(B, T, self.nq, Vnq // self.nq)
+
+        return logits
+
+    @torch.no_grad()
+    def inference(
+        self,
+        prefix: torch.Tensor,
+        opts: SpeechLMInferenceOptions,
+        enc_seq: torch.Tensor = None,
+        suffix: torch.Tensor = None,
+    ):
+        # (1) initialization
+        cache, hooks = install_kv_cache_hook(self.decoders, {})
+
+        # (2) prefix forward
+        prefix = prefix.expand(opts.nbest, -1, -1)
+        suffix = suffix.expand(opts.nbest, -1, -1)
+        _ = self.encode(prefix[:, :-1], cache=cache)  # exclude modality start
+
+        # (3) inference loop
+        minlen = int(prefix.size(1) * opts.minlenratio) if opts.minlenratio > 0 else 0
+        maxlen = int(prefix.size(1) * opts.maxlenratio)
+        if opts.search_algo == "teacher_force" and suffix is not None:
```

Comment:
Assertion to search algo would be necessary to keep it in the correct domain (not necessarily here, but it is necessary)